text/0000-projections-lifetimes-and-wf.md

```diff
@@ -0,0 +1,972 @@
+by looking at the relevant impl:
+
+```rust
+impl<'b,U> Iterator for Iter<'b,U> {
+    type Item = &'b U;
+    ...
+}
+```
+
+From this impl, we can conclude that `<PROJ> == &'a T`, and thus
+reduce `<PROJ>: 'x` to `&'a T: 'x`, which in turn holds if `'a: 'x`
+and `T: 'x` (from the rule `OutlivesReference`).
+
+But often we are in a situation where we can't normalize the
+projection. What can we do then? The rule
+`OutlivesProjectionComponents` says that if we can conclude that every
+lifetime/type parameter `Pi` to the trait reference outlives `'x`,
+then we know that a projection from those parameters outlives `'x`. In
+our example, the trait reference is `<Iter<'a,T> as Iterator>`, so
+that means that if the type `Iter<'a,T>` outlives `'x`, then the
+projection `<PROJ>` outlives `'x`. Now, you can see that this
+trivially reduces to the same result as the normalization, since
+`Iter<'a,T>: 'x` holds if `'a: 'x` and `T: 'x` (from the rule
+`OutlivesNominalType`).
+
+OK, so we've seen that applying the rule
+`OutlivesProjectionComponents` comes up with the same result as
+normalizing (at least in this case), and that's a good sign. But what
+is the basis of the rule?
+
+The basis of the rule comes from reasoning about the impl that we used
+to do normalization. Let's consider that impl again, but this time
+hide the actual type that was specified:
+
+```rust
+impl<'b,U> Iterator for Iter<'b,U> {
+    type Item = /* <TYPE> */;
+    ...
+}
+```
+
+So when we normalize `<PROJ>`, we obtain the result by applying some
+substitution `Θ` to `<TYPE>`. This substitution is a mapping from the
+lifetime/type parameters on the impl to some specific values, such
+that `<PROJ> == Θ <Iter<'b,U> as Iterator>::Item`. In this case, that
+means `Θ` would be `['b => 'a, U => T]` (and of course `<TYPE>` would
+be `&'b U`, but we're not supposed to rely on that).
+
+The key idea for the `OutlivesProjectionComponents` is that the only
+way that `<TYPE>` can *fail* to outlive `'x` is if either:
+
+- it names some lifetime parameter `'p` where `'p: 'x` does not hold; or,
+- it names some type parameter `X` where `X: 'x` does not hold.
+
+Now, the only way that `<TYPE>` can refer to a parameter `P` is if it
+is brought in by the substitution `Θ`. So, if we can just show that
+all the types/lifetimes that in the range of `Θ` outlive `'x`, then we
+know that `Θ <TYPE>` outlives `'x`.
+
+Put yet another way: imagine that you have an impl with *no
+parameters*, like:
+
+```rust
+impl Iterator for Foo {
+    type Item = /* <TYPE> */;
+}
+```
+
+Clearly, whatever `<TYPE>` is, it can only refer to the lifetime
+`'static`.  So clearly `<Foo as Iterator>::Item: 'static` holds. We
+know this is true without ever knowing what `<TYPE>` is -- we just
+need to see that the trait reference `<Foo as Iterator>` doesn't have
+any lifetimes or type parameters in it, and hence the impl cannot
+refer to any lifetime or type parameters.
```

Comment:
@Gankro 

> Something can only fail to outlive 'x if its type refers to (non-'static) lifetime that doesn't outlive 'x, and the only lifetimes it can possibly refer to have to be declared in the impl declaration. So it's sufficient (though perhaps not always necessary?) to make sure everything referred to in the impl declaration outlives 'x.

Yes, with a bit of subtlety -- we have to be sure that it not only appears in the impl declaration, but does not appear as part of a projection (as specified below). But this rule already exists.

In other words:

```
impl<'a> Trait for <i32 as SomeTrait<'a>>::Bar { ... }
```

wouldn't be good enough for `'a` because maybe `<i32 as SomeTrait<'a>>::Bar` normalized to `u32` (and `'a` doesn't appear in `u32`).